stdlib/public/runtime/SwiftRT-ELF.cpp

```diff
@@ -0,0 +1,61 @@
+//===--- SwiftRT-ELF.cpp --------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "ImageInspectionELF.h"
+
+#include <cstddef>
+
+// Create empty sections to ensure that the start/stop symbols are synthesized
+// by the linker.  Otherwise, we may end up with undefined symbol references as
+// the linker table section was never constructed.
+
+#define DECLARE_SWIFT_SECTION(name)                                            \
+  __asm__("\t.section " #name ",\"a\"\n");                                     \
+  extern __attribute__((__visibility__("hidden"))) void *__start_##name;       \
+  extern __attribute__((__visibility__("hidden"))) void *__stop_##name;
+
+DECLARE_SWIFT_SECTION(swift2_protocol_conformances)
+DECLARE_SWIFT_SECTION(swift2_type_metadata)
+
+DECLARE_SWIFT_SECTION(swift3_typeref)
+DECLARE_SWIFT_SECTION(swift3_reflstr)
+DECLARE_SWIFT_SECTION(swift3_fieldmd)
+DECLARE_SWIFT_SECTION(swift3_assocty)
+
+#undef DECLARE_SWIFT_SECTION
+
+extern "C" {
+__attribute__((__constructor__))
+static void swift_section_constructor() {
+#define SWIFT_SECTION_RANGE(name)                                              \
+  { reinterpret_cast<uintptr_t>(&__start_##name),                              \
+    reinterpret_cast<uintptr_t>(&__stop_##name) - reinterpret_cast<uintptr_t>(&__start_##name) }
+
+  const swift::MetadataSections sections = {
+    swift::CurrentSectionMetadataVersion,
+    0,
+
+    SWIFT_SECTION_RANGE(swift2_protocol_conformances),
+    SWIFT_SECTION_RANGE(swift2_type_metadata),
+
+    SWIFT_SECTION_RANGE(swift3_typeref),
+    SWIFT_SECTION_RANGE(swift3_reflstr),
+    SWIFT_SECTION_RANGE(swift3_fieldmd),
+    SWIFT_SECTION_RANGE(swift3_assocty),
+  };
+
+  swift_addNewDSOImage(&sections);
+
+#undef SWIFT_SECTION_RANGE
```

Comment:
This is doing quite a bit more work and hardcoding more ABI liability into the static constructor than before. The static constructor should ideally do nothing more than make a runtime call, so that the runtime can be upgraded with additional capabilities without recompiling dynamic libraries. Since the static constructor runs eagerly at load time for libraries the executable links against, it's also important that it do as little work as possible if no lazily-instantiated runtime mechanisms have been initialized that require runtime instantiation. I think the previous approach of having the runtime inspect the image is preferable for these reasons.

Reply:
I disagree. The new static constructor is doing negligible work: it merely collects a small number of image-local addresses and calls the runtime function. This arrangement avoids the vastly more expensive cost of calling `dlsym()` to find the sections later.

The version field in the `MetadataSections` provides sufficient ABI flexibility. New fields can be added to it; new runtimes can recognize calls from old libraries that don't have those fields.

Reply:
OTOH, this design means that future runtimes *only* have access to the information we have the forethought to feed them through this `MetadataSections` record, instead of potentially any information we can glean from the image. (One could argue the difference is marginal in practice, since we'd still need to consciously export a symbol to reliably discover anything dynamically in the old model anyway.)

It seems to me like we still shouldn't *have* to do any work in the static constructor here, since the data structure here is nothing but a bunch of local references. We should be able to make the `MetadataSections` record here a global constant, full of relative references, so the only load-time work the constructor has to do is call the runtime with a pointer to the global constant. Using a global constant could still give us the flexibility to `dladdr` the pointer and get back to a general dso handle we can do deeper inspection on if we wanted to in the future.

Reply:
@jckarter I'd *love* to make that a global constant.  I just could not figure out how to make that a global constant (at least within C).  I suppose that a C++ constructor could be emitted to construct the global structure.

IMO, this is doing a very small amount of work, which would ideally be just the runtime registration call, but I could not make the structure a global constant.  That would be my preference as well.  But, yes, the should is exactly what I think is ideal: the structure is a constant global, the constructor is just to tie to into the runtime.

Reply:
Yeah, neither C's nor C++'s model for constant expressions is expressive enough here. You might be able to express it as LLVM IR, though, or as global inline assembly. Since it's just data it should be portable.

Reply:
Calling `dladdr()` on any of the values passed in `MetadataSections` ought to recover the dso handle. We could add a field to `MetadataSections` that is simply a non-nil pointer somewhere into the image, to ensure that `dladdr()` has a target even if the image happens to have none of the other data.

Reply:
@gparker42 that is one of the tricks that I'm playing here.  The reason for the global inline assembly, as gross as it may be, is that it forces that the section will be constructed, which will force the emission of the start/stop symbols.  These are referenced by the table thus preventing the dead strip from kicking in.  As a result, we are guaranteed that the markers will always be present, though the section may be 0-sized.  The symbols are image specific, so we are guaranteed to have a local symbol to reference to get the DSO handle.